Adaptors/LocalStorage/src/ObjectStorage.cs

```diff
@@ -0,0 +1,168 @@
+                       int                    chunkSize,
+                       ILogger<ObjectStorage> logger)
+  {
+    path_      = path;
+    logger_    = logger;
+    chunkSize_ = chunkSize;
+
+    logger.LogDebug("Creating Local ObjectStorage at {path}",
+                    path);
+
+    Directory.CreateDirectory(path);
+  }
+
+  /// <inheritdoc />
+  public Task AddOrUpdateAsync(string                   key,
+                               IAsyncEnumerable<byte[]> valueChunks,
+                               CancellationToken        cancellationToken = default)
+    => AddOrUpdateAsync(key,
+                        valueChunks.Select(chunk => (ReadOnlyMemory<byte>)chunk.AsMemory()),
+                        cancellationToken);
+
+  /// <inheritdoc />
+  public async Task AddOrUpdateAsync(string                                 key,
+                                     IAsyncEnumerable<ReadOnlyMemory<byte>> valueChunks,
+                                     CancellationToken                      cancellationToken = default)
+  {
+    var filename = Path.Combine(path_,
+                                key);
+
+    using var _ = logger_.LogFunction(filename);
+
+    // Write to temporary file
+    await using var file = File.Open(filename,
+                                     FileMode.OpenOrCreate,
+                                     FileAccess.Write);
+
+    // TODO: Overlap chunk reading and chunk writing
+    await foreach (var chunk in valueChunks.WithCancellation(cancellationToken)
+                                           .ConfigureAwait(false))
+    {
+      await file.WriteAsync(chunk,
+                            cancellationToken)
+                .ConfigureAwait(false);
+    }
+
+    await file.FlushAsync(cancellationToken)
+              .ConfigureAwait(false);
+  }
+
+  /// <inheritdoc />
+  public async IAsyncEnumerable<byte[]> GetValuesAsync(string                                     key,
+                                                       [EnumeratorCancellation] CancellationToken cancellationToken = default)
+  {
+    var filename = Path.Combine(path_,
+                                key);
+
+    using var _ = logger_.LogFunction(filename);
+
+    if (!File.Exists(filename))
+    {
+      throw new ObjectDataNotFoundException($"The object {key} has not been found in {path_}");
+    }
+
+    await using var file = File.Open(filename,
+                                     FileMode.Open,
+                                     FileAccess.Read);
+
+    // Task is not awaited here in order to overlap reading and yielding
+    var buffer = new byte[chunkSize_];
+    var readTask = file.ReadAsync(buffer,
+                                  cancellationToken);
+    while (true)
+    {
+      // We need to await the read task before starting to read another chunk
+      var read       = await readTask.ConfigureAwait(false);
+      var readBuffer = buffer;
+
+      if (read == 0)
+      {
+        yield break;
+      }
+
```

Comment:
you should be able to put that in the while

Reply:
I don't a see way to do it without duping most of the loop body.

Reply:
```suggestion
    int read;
    while ((read = await readTask.ConfigureAwait(false)) == 0)
    {
      var readBuffer = buffer;

```

Reply:
Moreover, you can do things after the `yield return` but before the end of the while block.